File: 2006/impl/java/src/um/MemoryManager.java
/*
 * MemoryManager.java
 *
 * Created on 25 Октябрь 2006 г., 13:37
 *
 * To change this template, choose Tools | Template Manager
 * and open the template in the editor.
 */

package um;

import java.util.ArrayList;
import java.util.List;

/**
 *
 * @author lattyf
 */
public class MemoryManager
{
    
    private static MemoryManager instance;
    private List<long[]> arrays;
    
    /** Creates a new instance of MemoryManager */
    private MemoryManager()
    {
        arrays = new ArrayList<long[]>();
        arrays.add(null);
    }
    
    public static MemoryManager getInstance()
    {
        if (instance == null)
            instance = new MemoryManager();
        
        return instance;
    }
    
    public int allocate(int aCapasity)
    {
        long[] newArray = new long[aCapasity];
        
        for (int i = 0; i < aCapasity; ++i)
            newArray[i] = 0;
        
        int newArrayID = getFreeArrayID();
        if (arrays.size() == newArrayID)
        	arrays.add(null);
        
        arrays.set(newArrayID, newArray);
        return newArrayID;
    }
    
    public void abandon(int anArrayID)
    {
        arrays.set(anArrayID, null);
    }
    
    public List<long[]> getArrays()
    {
        return arrays;
    }
    
    private int getFreeArrayID()
    {
        int arraysSize = arrays.size();
        
        for (int i = 0; i < arraysSize; ++i)
            if (arrays.get(i) == null)
            {
                return i;
            }
        
        return arraysSize;
    }
    
    public void copyToZeroArray(int anArrayID)
    {
        if (anArrayID == 0)
            return;
        
        if (arrays.get(0) != null)
        {
            arrays.set(0, null);
        }
        
        arrays.set(0, (long[]) arrays.get(anArrayID).clone());
    }
    
    public void loadScroll(long[] aScroll)
    {
        if (arrays.get(0) != null)
        {
            arrays.set(0, null);
        }
        
        arrays.set(0, (long[]) aScroll.clone());
    }
    
    
}

File: 2006/impl/mono/Command/MemoryManager.cs
using System;
using System.Collections.Generic;

namespace UM.Command
{
    public class MemoryManager
    {
        private static MemoryManager instance;
        private IList<uint[]> arrays;

        public static MemoryManager GetInstance()
        {
            if (instance == null)
                instance = new MemoryManager();

            return instance;
        }

        public MemoryManager()
        {
            arrays = new List<uint[]> {null};
        }

        public uint Allocate(uint aCapasity)
        {
            uint[] newArray = new uint[aCapasity];

            for (int i = 0; i < aCapasity; ++i)
                newArray[i] = 0;

            uint newArrayID = getFreeArrayID();
            if (arrays.Count == newArrayID)
                arrays.Add(null);

            arrays[(int) newArrayID] = newArray;
            return newArrayID;
        }

        public void Abandon(uint anArrayID)
        {
            arrays[(int) anArrayID] = null;
        }

        private uint getFreeArrayID()
        {
            uint arraysSize = (uint) arrays.Count;

            for (uint i = 0; i < arraysSize; ++i)
                if (arrays[(int) i] == null)
                {
                    return i;
                }

            return arraysSize;
        }

        public void CopyToZeroArray(uint anArrayID)
        {
            if (anArrayID == 0)
                return;

            if (arrays[0] != null)
            {
                arrays[0] = null;
            }

            arrays[0] = new uint[arrays[(int) anArrayID].Length];
            Array.Copy(arrays[(int) anArrayID], arrays[0], arrays[(int) anArrayID].Length);
        }

        public void LoadScroll(uint[] aScroll)
        {
            if (arrays[0] != null)
            {
                arrays[0] = null;
            }

            arrays[0] = new uint[aScroll.Length];
            Array.Copy(aScroll, arrays[0], aScroll.Length);
        }

        public uint this[uint i, uint currentOffset]
        {
            get { return arrays[(int)i][currentOffset]; }
            set { arrays[(int)i][currentOffset] = value; }
        }
    }
}